### PR TITLE
Make Type[X] compatible with metaclass of X

### LINKED ISSUE #3326: Type[X] not compatible with metaclass of X
The second-to-last line should probably be accepted by mypy, but it currently generates an error because `Type[A]` isn't compatible with the metaclass of `M`:

```py
from typing import Type

class M(type): pass
class A(metaclass=M): pass

a: Type[A] = A
m: M = a    # error, but should be okay
m2: M = A   # no error
```

### PR DESCRIPTION
**Pull request by elazarg**
Fix #3326.

PR #2837 seems completely broken; the checks were all backwards, so this PR reverses them.

I had to special-case EnumMeta since the bound for the self argument can't be lower than the current class. I fail to see how it can work without something like `metaclass=M` argument for TypeVar.

### CODE REVIEW

File: mypy/subtypes.py [hunk 1]
```diff
@@ -155,19 +150,18 @@ def visit_instance(self, left: Instance) -> bool:
                        for lefta, righta, tvar in
                        zip(t.args, right.args, right.type.defn.type_vars))
         if isinstance(right, TypeType):
-            item = right.item
-            if isinstance(item, TupleType):
-                item = item.fallback
```
Comment (ilevkivskyi) [inline mypy/subtypes.py]: Why did you remove the case of ``TupleType``? A named tuple can appear in ``Type[]``.

File: mypy/subtypes.py [hunk 2]
```diff
@@ -263,8 +257,7 @@ def visit_overloaded(self, left: Overloaded) -> bool:
         elif isinstance(right, TypeType):
             # All the items must have the same type object status, so
             # it's sufficient to query only (any) one of them.
-            # This is unsound, we don't check the __init__ signature.
```
Comment (ilevkivskyi) [inline mypy/subtypes.py]: I think this comment should stay here, since the ``__init__`` signature should be checked for all items in an overload.

File: test-data/unit/check-classes.test [hunk 3]
```diff
@@ -3066,12 +3066,13 @@ class M(type):
 class A(metaclass=M): pass
 
 reveal_type(A[M])  # E: Revealed type is 'builtins.int'
+reveal_type(A[M])  # E: Revealed type is 'builtins.int'
```
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: This reveal is identical to the previous one. I think you wanted to check something like ``A[A]``.

File: test-data/unit/check-classes.test [hunk 4]
```diff
@@ -3137,3 +3138,51 @@ class M(type):
...
 
+[case testMetaclassStrictSupertypeOfTypeWithClassmethods]
+from typing import Type, TypeVar
+TA = TypeVar('TA', bound='A')
+TTA = TypeVar('TTA', bound='Type[A]')
+TM = TypeVar('TM', bound='M')
+
+class M(type):
+    def g1(cls: 'Type[A]') -> None: pass #  E: The erased type of self 'Type[__main__.A]' is not a supertype of its class '__main__.M'
+    def g2(cls: Type[TA]) -> None: pass #  E: The erased type of self 'Type[__main__.A]' is not a supertype of its class '__main__.M'
```
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: Although ``Type[A]`` is lower than ``M``, I think this could be allowed. The special-casing probably should be in ``checkmember.bind_self`` (then you will also not need special-casing ``EnumMeta`` above). My idea is following:
```python
class M(type):
    def g2(cls: Type[TA]) -> TA:
        ...
    def g4(cls: TM) -> None:
        ...
class A(metaclass=M): ...
class A2(A): ...
class B(metaclass=M): ...
A.g2() # OK, inferred type is 'A'
A2.g2() # OK, inferred type is 'A2'
B.g2() # Error, g2 should be only called on subclasses of A
A.g4(), A2.g4(), B.g4() # All are OK
```
The situation with one "privileged" class with a given metaclass sometimes appears in frameworks. This will allow users to give more precise types in such situation.

File: test-data/unit/check-classes.test [hunk 5]
```diff
@@ -3137,3 +3138,51 @@ class M(type):
...
+    def g3(cls: TTA) -> None: pass #  E: The erased type of self 'Type[__main__.A]' is not a supertype of its class '__main__.M'
+    def g4(cls: TM) -> None: pass
+m: M
+
+class A(metaclass=M):
+    def foo(self): pass
+
+# 3 examples of unsoundness - instantiation, classmethod, staticmethod and ClassVar:
```
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: It looks here are _four_ examples.

File: test-data/unit/check-classes.test [hunk 6]
```diff
@@ -3137,3 +3138,51 @@ class M(type):
...
+x2: M = s
+
+from typing import ClassVar
+class Cvar(metaclass=M):
+    x = 1  # type: ClassVar[int]
+cv: Type[Cvar] = m  # E: Incompatible types in assignment (expression has type "M", variable has type Type[Cvar])
+cv.x
+x3: M = cv
+
+[builtins fixtures/classmethod.pyi]
```
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: I would add few ``reveal_type``s for methods in these tests, just in case.

File: test-data/unit/check-classes.test (hunk 3 above)
Comment (elazarg) [inline test-data/unit/check-classes.test]: Nah I've probably copied it with the test below somehow :)

File: test-data/unit/check-classes.test (hunk 4 above)
Comment (elazarg) [inline test-data/unit/check-classes.test]: So this is probably a more general decision to be made. Instead of requiring a supertype, we can require a related type - either a supertype or a subtype, in which case there's a check at the point of use, just like any other function.

Comment (elazarg) [inline test-data/unit/check-classes.test]: Although I'm not sure why add `g2` in M instead of in `A` as a classmethod

File: test-data/unit/check-classes.test (hunk 6 above)
Comment (elazarg) [inline test-data/unit/check-classes.test]: I tried, although I'm not sure what information is gained.

Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: Actually I wanted something like ``reveal_type(A.g1)`` and ``reveal_type(A.g1())``, in particular, where return types of ``g2`` etc. depend on type variables (like ``List[TA]``). Plus also the situations where name of staticmethod/classmethod in class coincides with a name of a method in metaclass.

File: test-data/unit/check-classes.test (hunk 4 above)
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: > So this is probably a more general decision to be made. Instead of requiring a supertype, we can require a related type - either a supertype or a subtype, in which case there's a check at the point of use, just like any other function.

I would vote for doing this first only for self types in metaclasses (since this is a situation where we know this is needed).

> Although I'm not sure why add ``g2`` in M instead of in ``A`` as a classmethod

Yes, but this will not work for dunder methods.

Comment (elazarg) [inline test-data/unit/check-classes.test]: I see.
Adding checks at access seems to have non local effect, which makes this PR significantly more complex. Perhaps it will be better to split it up, and add the check at access only later.

File: test-data/unit/check-classes.test (hunk 6 above)
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: > Actually I wanted something like ``reveal_type(A.g1)`` and ``reveal_type(A.g1())``, in particular, where return types of ``g2`` etc. depend on type variables (like ``List[TA]``). Plus also the situations where name of staticmethod/classmethod in class coincides with a name of a method in metaclass.

@elazarg Have you added these tests? Otherwise I think this PR looks good.
@JukkaL could you please try this PR with the Dropbox internal codebase before it is merged?

Comment (elazarg) [inline test-data/unit/check-classes.test]: Hmm I'm pretty sure I did, but probably forgot to push... :\ I will check.

Comment (elazarg) [inline test-data/unit/check-classes.test]: Ok, I don't have time at this moment, so I tried pushing kinda blindly. I might have mixed up something. This supposed to be the tests only, not the check-on-access you've asked for. I hope I did not mess up too bad; if I did, I can probably try to fix it later this week. In such case, it can still be useful to check earlier commit against the internal repo.

Sorry about the mess.

Comment (JukkaL) [inline test-data/unit/check-classes.test]: @ilevkivskyi I can try this with Dropbox internal code tomorrow (need to go now).

Comment (gvanrossum) [inline test-data/unit/check-classes.test]: No worries, I have tried this with our internal code, there are no new errors this found (either in our codebase or in the PR). I haven't reviewed the PR though so I can't "Approve" it. But Ivan can.
